### PR TITLE
#6403 Default Parameter Code Valid Value Value Corrections

### LINKED ISSUE #6403: Default Parameter Code Valid Value Value Corrections
## Overview

The DTO check for valid Default Parameter Codes values excludes "SORX" (Maximum SO2 Emission Rate).  The Monitor Default edit screens Parameter dropdown does not include "Maximum SO2 Emission Rate" (SORX).  Both imports and the screen should allow the use of "SORX".

The DTO check for valid Default Parameter Codes values excludes "SORX" (Maximum SO2 Emission Rate) and includes "BWA" (Moisture Fraction in Ambient Air).  Twenty-three instances of Default Parameter Code "SORX" currently exist with all beginning on January 1<sup>st</sup>, 2009.  No instances of "BWA" exists and could only be entered through a screen in ECMPS 1.0.  The purpose of this ticket is to update the DTO check to allow the import and entry of MP JSON that contains "SORX" Default Parameter Codes.

### Background

[Issues Found During Performance Testing](https://github.com/US-EPA-CAMD/easey-ui/issues/6403#issuecomment-2364657273)

## Changes

Make the following generic crosscheck changes to:

1. Allow the **DTO checking** for valid **Default Parameter Codes** to include "**SORX**".
2. Allow the creation, editing, and display of "**SORX**" defaults in the Monitor Default create/edit screen.
3. Correct the creation, editing, and display of "**BWA**" defaults in the Monitor Default create/edit screen.

| Id | Name | Value 1 | Value 2 |
| ---: | :--- | :--- | :--- |
| 11 | Parameter to Category | SORX | DEFAULT<sup>1</sup> |
| 191 | Parameter Code to Units of Measure Code for Defaults | BWA | PCT<sup>2</sup> |
| 191 | Parameter Code to Units of Measure Code for Defaults | SORX | LBMMBTU |
| 192 | Parameter Code to Purpose Code for Defaults | BWA | PM |
| 192 | Parameter Code to Purpose Code for Defaults | SORX | MD |
| 193 | Parameter Code to Fuel Code for Defaults | BWA | [see note]<sup>3</sup> |
| 193 | Parameter Code to Fuel Code for Defaults | SORX | PNG |
| 193 | Parameter Code to Fuel Code for Defaults | SORX | NFS |
| 193 | Parameter Code to Fuel Code for Defaults | SORX | DSL |
| 194 | Parameter Code to Source Code for Defaults | BWA | DATA |
| 194 | Parameter Code to Source Code for Defaults | SORX | DATA |
| 194 | Parameter Code to Source Code for Defaults | SORX | DEF |

#### Notes:

1. Update the existing "Parameter to Category" "SORX" Value 2 to "DEFAULT".
2. CAPD confirmed (Support Meeting October 10<sup>th</sup> 2024) to use "PCT" (Percentage) as the UOM for BWA in ECMPS 2.0, Any changes to this will be handled by a subsequent discussion and ticket.
   - CAPD confirmed this in the October 10<sup>th</sup> 2024 ECMPS Support Meeting.
   - **The ECMPS 2.0 Reporting Instructions may need to be updated to match this change**.
3. Add a line for each Fuel Code in the CAMDECMPSMD FUEL_CODE table.
   - CAPD confirmed this in the October 10<sup>th</sup> 2024 ECMPS Support Meeting.
   - **The ECMPS 2.0 Reporting Instructions may need to be updated to match this change**.
4. Value 3 is null for each row.
5. The changes will affect the data returned by CAMDECMPSMD.VW_DEFAULTS_MASTER_DATA_RELATIONSHIPS, which is used by the DTO checks and to populate the Monitor Default screen dropdowns.

## Technical Information

Currently the DTO check for the Default Parameter Code uses the following to get a list of the valid parameter codes.

`SELECT distinct parameter_code as "value" FROM camdecmpsmd.vw_defaults_master_data_relationships

SELECT * FROM camdecmpsmd.vw_defaults_master_data_relationships

> It returns "**BWA**" but does not return "**SORX**".  "**BWA**" is not a Default Parameter Code in the **ECMPS 1.0 XSD**'s **DefaultParameterCodeType**, but "**SORX**" is.

### Locations with "SORX" Default Parameter Codes

| ORIS | Facility | Locations |
| ---: | :--- | :--- |
| 1913 | Inver Hills | 1, 2, 3, 4, 5, 6 |
| 4014 | Wheaton Generating Plant | 1, 2, 3, 4, 5, 6 |
| 10143 | Colver Green Energy | AAB01 |

## Acceptance Tests:
1. Allow importing of "SORX" and "BWA" defaults. (Currently there are no BWA defaults)
2. Allow the creation, editing, and display of "SORX" defaults in the Monitor Default create/edit screen.
3. Correct the creation, editing, and display of "BWA" defaults in the Monitor Default create/edit screen.
4. Evaluate the data and make sure no issues are encountered.

### PR DESCRIPTION
**Pull request by EvanSun220**
- Removed related logic to set existing parameter to category SORX value 2 to null
- Added the record of set PCT as the UOM(Unit of Measure) of BWA
- Added one record for each of the fuel code for "Parameter code to fuel code for defaults" with value1=BWA

### CODE REVIEW

**Review by ergjustin — COMMENTED** [review]
Did you compiled in dev?
You need to add deployment script.(deployments -> ecmps -> release-v2.0-fix -> sprint12 -> ticket number -> your changes)

example:  https://github.com/US-EPA-CAMD/easey-db-scripts/pull/34

File: deployments/ecmps/release-v2.0-fix/Sprint12/6403/defaults_master_data_relationships_load.sql [hunk 1]
```diff
@@ -0,0 +1,642 @@
...
+set value2=null
+WHERE cross_chk_catalog_value_id=(select cross_chk_catalog_value_id
+									from camdecmpsmd.cross_check_catalog_value 
+									where cross_chk_catalog_id=(select cross_chk_catalog_id 
+																from camdecmpsmd.cross_check_catalog
+																where cross_chk_catalog_name='Parameter to Category') 
+																	and value1='AKSF' 
+																	and value2='DEFAULT');
+
+-- Update all the existing "Parameter to Category" "SORX" Value 2 to "DEFAULT".
```
Comment (djw4erg) [inline deployments/ecmps/release-v2.0-fix/Sprint12/6403/defaults_master_data_relationships_load.sql:12]: **Just a note**, and I realize you used previous SQL as the model, but "camdecmpsmd.cross_check_catalog_value" being the main table and also used in the where clause is very odd.  **In the future** I would use something like the following:

```
update  camdecmpsmd.cross_check_catalog_value 
   set  value2 = 'DEFAULT'
 where  cross_chk_catalog_id = (
                                select  cross_chk_catalog_id
                                  from  camdecmpsmd.cross_check_catalog
                                 where  cross_chk_catalog_name = 'Parameter to Category')
                               )
   and  value1 = 'SORX';
```

Comment (EvanSun220) [reply deployments/ecmps/release-v2.0-fix/Sprint12/6403/defaults_master_data_relationships_load.sql:12]: Thanks for the reminder Dwayne!

File: deployments/ecmps/release-v2.0/mdm-relationships/defaults_master_data_relationships_load.sql
Comment (djw4erg) [inline deployments/ecmps/release-v2.0/mdm-relationships/defaults_master_data_relationships_load.sql:1]: #### All of the "**SORX**" updates are missing except for the "**Parameter to Category**" (Id: 11) update.

There may have been some confusion with the existing "**SO2R**" and "**SO2X**" entries.

| Id | Name | Value 1 | Value 2 |
| ---: | :--- | :--- | :--- |
| 191 | Parameter Code to Units of Measure Code for Defaults | SORX | LBMMBTU |
| 192 | Parameter Code to Purpose Code for Defaults | SORX | MD |
| 193 | Parameter Code to Fuel Code for Defaults | SORX | DSL |
| 193 | Parameter Code to Fuel Code for Defaults | SORX | NFS |
| 193 | Parameter Code to Fuel Code for Defaults | SORX | PNG |
| 194 | Parameter Code to Source Code for Defaults | SORX | DATA |
| 194 | Parameter Code to Source Code for Defaults | SORX | DEF |

**Review by djw4erg — CHANGES_REQUESTED** [review]

File: deployments/ecmps/release-v2.0/mdm-relationships/defaults_master_data_relationships_load.sql
Comment (EvanSun220) [reply deployments/ecmps/release-v2.0/mdm-relationships/defaults_master_data_relationships_load.sql:1]: Hi Dwayne, all of the missed SORX related records are added and built in dev db.

File: deployments/ecmps/release-v2.0/mdm-relationships/defaults_master_data_relationships_load.sql [hunk 2]
```diff
@@ -521,7 +524,41 @@ INSERT INTO camdecmpsmd.cross_check_catalog_value(cross_chk_catalog_id, value1,
...
+INSERT INTO camdecmpsmd.cross_check_catalog_value(cross_chk_catalog_id, value1, value2, value3) VALUES ((select cross_chk_catalog_id from camdecmpsmd.cross_check_catalog where cross_chk_catalog_name='Parameter Code to Fuel Code for Defaults'), 'BWA', 'TDF', null);
+INSERT INTO camdecmpsmd.cross_check_catalog_value(cross_chk_catalog_id, value1, value2, value3) VALUES ((select cross_chk_catalog_id from camdecmpsmd.cross_check_catalog where cross_chk_catalog_name='Parameter Code to Fuel Code for Defaults'), 'BWA', 'W', null);
+INSERT INTO camdecmpsmd.cross_check_catalog_value(cross_chk_catalog_id, value1, value2, value3) VALUES ((select cross_chk_catalog_id from camdecmpsmd.cross_check_catalog where cross_chk_catalog_name='Parameter Code to Fuel Code for Defaults'), 'BWA', 'WF', null);
+INSERT INTO camdecmpsmd.cross_check_catalog_value(cross_chk_catalog_id, value1, value2, value3) VALUES ((select cross_chk_catalog_id from camdecmpsmd.cross_check_catalog where cross_chk_catalog_name='Parameter Code to Fuel Code for Defaults'), 'SORX', 'DSL', null);
```
Comment (djw4erg) [inline deployments/ecmps/release-v2.0/mdm-relationships/defaults_master_data_relationships_load.sql:559]: I am not sure how, but the following now appears in the database twice.  **The script, however, looks good.**

| Id | Name | Value 1 | Value 2 |
| ---: | :--- | :--- | :--- |
| 193 | Parameter Code to Fuel Code for Defaults | SORX | DSL |

**Review by djw4erg — CHANGES_REQUESTED** [review]

File: deployments/ecmps/release-v2.0/mdm-relationships/defaults_master_data_relationships_load.sql (hunk 2 above)
Comment (EvanSun220) [reply deployments/ecmps/release-v2.0/mdm-relationships/defaults_master_data_relationships_load.sql:559]: Hi Dwayne, the issue is resolved. Please check if everything is as expected.

**Review by djw4erg — APPROVED** [review]

**Pull request merged**